core/src/main/java/org/elasticsearch/index/mapper/IdFieldMapper.java

```diff
@@ -113,32 +105,64 @@ public boolean isSearchable() {
 
         @Override
         public Query termQuery(Object value, @Nullable QueryShardContext context) {
-            final BytesRef[] uids = Uid.createUidsForTypesAndId(context.queryTypes(), value);
-            return new TermInSetQuery(UidFieldMapper.NAME, uids);
+            return termsQuery(Arrays.asList(value), context);
         }
 
         @Override
-        public Query termsQuery(List values, @Nullable QueryShardContext context) {
+        public Query termsQuery(List<?> values, @Nullable QueryShardContext context) {
+            if (indexOptions() != IndexOptions.NONE) {
+                return super.termsQuery(values, context);
+            }
             return new TermInSetQuery(UidFieldMapper.NAME, Uid.createUidsForTypesAndIds(context.queryTypes(), values));
         }
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder() {
+            if (indexOptions() == IndexOptions.NONE) {
+                throw new IllegalArgumentException("Fielddata access on the _uid field is disallowed");
+            }
+            return new PagedBytesIndexFieldData.Builder(
+                    TextFieldMapper.Defaults.FIELDDATA_MIN_FREQUENCY,
+                    TextFieldMapper.Defaults.FIELDDATA_MAX_FREQUENCY,
+                    TextFieldMapper.Defaults.FIELDDATA_MIN_SEGMENT_SIZE);
+        }
+    }
+
+    static MappedFieldType defaultFieldType(Settings indexSettings) {
```

Comment:
you have `IndexSettings` available when you create `IdFieldMapper` I wonder if we can just pass a boolean down to the ctor if we have more than one type?